ci(wasm): add GHA to deploy wasm playground previews on PRs

gcharang opened the pull request.

gcharang committed:
adds gh action to deploy wasm playground previews on PRs

smk762 reviewed (commented):
Love it. Version in header is useful to be sure testing on correct branch.  
One issue which affects some methods: 

<img width="753" height="631" alt="image" src="https://github.com/user-attachments/assets/23016964-30ae-4b30-b873-e6d7f91c93e0" />

I think this `name` param is internal for the list making, can we please `del` it on the way to the request panel? Some methods dont care if its there, others are not so forgiving.

In future, we could default it to the postman collection in branch being deployed (pending other automations).

smk762 commented:
optional enhancement idea: In https://github.com/KomodoPlatform/komodo-docs-mdx/pull/572/files#diff-459a511b5145b027153cef692d0071ecb1cbdab37c10eb1156d666869bfb385fR113-R116 the workflow run is conditional upon the presence of a label within the PR. 

With this, we can limit deployments only to PRs which have signalled they are ready for and/or require deployment (e.g. dont deploy just because README.md updated).

mariocynicys commented on .github/workflows/pr-wasm-preview.yml:
```diff
@@ -0,0 +1,204 @@
+      
+      - name: Cache yarn dependencies
+        uses: actions/cache@v4
+        id: yarn-cache
```
guess there is no need to id this cache step

mariocynicys commented on Cargo.toml:
```diff
@@ -236,7 +236,7 @@ debug-assertions = false
 strip = true
 codegen-units = 1
-# lto = true
+lto = true
```
what are these changes for? what they do?

onur-ozkan replied:
Please revert this change. KDF builds are already taking extremely long time, this will at least triple it.

onur-ozkan replied:
> what they do?

It optimizes the whole binary on link time (the last step of the compilation).

gcharang replied:
done. reverted this change

mariocynicys commented on mm2src/mm2_bin_lib/Cargo.toml:
```diff
@@ -62,4 +62,4 @@ jemallocator.workspace = true
 
 [package.metadata.wasm-pack.profile.release]
-wasm-opt = ['-O', '--enable-bulk-memory', '--enable-nontrapping-float-to-int']
+wasm-opt = ['-Oz', '--enable-bulk-memory', '--enable-nontrapping-float-to-int', '--strip-debug']
```
this too

gcharang replied:
this and lto = true were my attempt to make the wasm bin smaller. cloudflare pages has a max file size limit of 25mb but the wasm file is more than 30 mb. I solved it by gzipping the wasm lib and publishing it. user's browser will decompress it and use it

onur-ozkan replied:
Dropping debuginfo on playground environments is fishy. We should also avoid adding LTO. Can we deploy them somewhere else than cloudflare?

onur-ozkan replied:
The size is also dynamic and be still larger than 25MB in the future even with dropping the debuginfo. If we can deploy them at some other server that doesn't give small limits that would be perfect.

gcharang replied:
I've solved the size issue for now by gzipping the wasm library and decompressing it for use on the client browser. will revert the optimization related changes I did as they are unnecessary now

gcharang replied:
(Original WASM: 33M, Gzipped WASM: 11M) in the deployement links comment https://github.com/KomodoPlatform/komodo-defi-framework/pull/2607#issuecomment-3215990383 tracks this

gcharang replied:
done. reverted this change

mariocynicys reviewed (commented):
have some questions but lgtm otherwise.
the `name` thing @smk762 mentioned i think is to be fixed from `react-komodefi-wasm`, so that's not a blocker here i guess.

gcharang commented:
> optional enhancement idea: In [KomodoPlatform/komodo-docs-mdx#572 (files)](https://github.com/KomodoPlatform/komodo-docs-mdx/pull/572/files#diff-459a511b5145b027153cef692d0071ecb1cbdab37c10eb1156d666869bfb385fR113-R116) the workflow run is conditional upon the presence of a label within the PR.
> 
> With this, we can limit deployments only to PRs which have signalled they are ready for and/or require deployment (e.g. dont deploy just because README.md updated).

I made the workflow run only when there are changes in src: [`787e653` (#2607)](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2607/commits/787e65350b2f6784257f64bbe233b9296162454f) so no need to check for a label right? even if someone forgets to add a label, if a file in mm2src was changed, the change probably needs a deployment
@smk762

smk762 commented:
> > optional enhancement idea: In [KomodoPlatform/komodo-docs-mdx#572 (files)](https://github.com/KomodoPlatform/komodo-docs-mdx/pull/572/files#diff-459a511b5145b027153cef692d0071ecb1cbdab37c10eb1156d666869bfb385fR113-R116) the workflow run is conditional upon the presence of a label within the PR.
> > With this, we can limit deployments only to PRs which have signalled they are ready for and/or require deployment (e.g. dont deploy just because README.md updated).
> 
> I made the workflow run only when there are changed in src: [`787e653` (#2607)](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2607/commits/787e65350b2f6784257f64bbe233b9296162454f) so no need to check for a label right? even if someone forgets to add a label, if a file in mm2src was changed, the change probably needs a deployment @smk762

I'll leave it to KDF team - as long as there is no quota/cost concern with deploying freely, its better to have than not. Idea for the filter was only to limit where PR incomplete/draft to avoid unused deployments.

onur-ozkan commented on .github/workflows/pr-wasm-preview.yml:
```diff
@@ -0,0 +1,210 @@
+      - synchronize
+      - reopened
+    paths:
+      - 'mm2src/**'
+      - Cargo.lock
+      - Cargo.toml
+      - rust-toolchain.toml
```
`.cargo` should also be included. If it doesn't costs us, we can just include the whole repo for safety (e.g., adding new directory and forgetting to add it here).

gcharang replied:
yes, removed path based check and added `    if: github.event.pull_request.labels[0] != null` conditional to make it run for pushes if a label is present for the PR

gcharang commented:
@smk762 removed the path checks and made the deployments run only if there's a label added to the PR

smk762 commented:
Confirmed https://github.com/KomodoPlatform/komodo-defi-wasm-rpc/issues/3 resolved in latest deployment and playground functional. Some nice value-adds with the tree and PR links :medal_sports:

shamardy commented:
> I'll leave it to KDF team - as long as there is no quota/cost concern with deploying freely, it's better to have than not. Idea for the filter was only to limit where PR incomplete/draft to avoid unused deployments.

> removed the path checks and made the deployments run only if there's a label added to the PR

I would prefer to deploy with a certain specified label not any label, even if a PR is in draft and the author wants to deploy a wasm playground to test some things they can place such label. This is the best solution cost wise, I will add such label and inform you of it.

shamardy commented:
I guess `deploy: wasm-playground` would be a good label

gcharang commented:
@shamardy should it be like:

if (label named 'deploy: wasm-playground' is present) { run the action independent of any other status } else {don't run action}

or

if (PR is not draft) { run action } if (PR is draft and label named 'deploy: wasm-playground' is present) {run action}

shamardy commented:
It should be `if (label named 'deploy: wasm-playground' is present) { run the action independent of any other status } else {don't run action}` but also trigger the action when the label is added not only on a new commit.

shamardy commented on .github/workflows/pr-wasm-preview.yml:
```diff
@@ -0,0 +1,206 @@
+          path: react-komodefi-wasm
+
+      - name: Setup Node.js
+        uses: actions/setup-node@v3
```
nit: Let's use v4 here similar to build cache action

gcharang replied:
done

shamardy commented on .github/workflows/pr-wasm-preview.yml:
```diff
@@ -0,0 +1,221 @@
+    runs-on: ubuntu-latest
+    container: komodoofficial/ci-container:latest
+    steps:
+      - name: Check label
+        id: check
+        run: |
+          has_label=$(jq -r '.pull_request.labels[].name' "$GITHUB_EVENT_PATH" | grep -x 'deploy: wasm-playground' || true)
+          if [ -n "$has_label" ]; then
+            echo "should_run=true" >> $GITHUB_OUTPUT
+          else
+            echo "should_run=false" >> $GITHUB_OUTPUT
+          fi
```
Prefer a job-level `if` that checks for the label, this way the job won’t even start unless the label is present which is cleaner than starting and exiting early.

gcharang replied:
done

shamardy commented on .github/workflows/pr-wasm-preview.yml:
```diff
@@ -0,0 +1,221 @@
+      - synchronize
+      - reopened
+      - labeled
+      - unlabeled
```
`unlabeled` shouldn't be needed if we used
```
if: contains(github.event.pull_request.labels.*.name, 'deploy: wasm-playground')
```
As suggested in the previous comment.

gcharang replied:
done

shamardy reviewed (commented):
Thanks for the fixes! I have a few comments :)

shamardy reviewed (dismissed):
🔥

onur-ozkan commented on .github/workflows/pr-wasm-preview.yml:
```diff
@@ -0,0 +1,207 @@
+      - labeled
+
+env:
+  JEMALLOC_SYS_WITH_MALLOC_CONF: "background_thread:true,narenas:1,tcache:false,dirty_decay_ms:0,muzzy_decay_ms:0,metadata_thp:auto"
```
Can you explain why do we need this env?

shamardy replied:
I guess it was copied from other CI build scripts, it's not needed for wasm / this usecase as we enable `jemallocator` for linux only.

shamardy replied:
ref. https://github.com/KomodoPlatform/komodo-defi-framework/blob/daf43ebeadf0b8e17f2b76019c6496236f4dd46c/mm2src/mm2_bin_lib/src/mm2_bin.rs#L10-L12

gcharang replied:
copy and pasted from here https://github.com/KomodoPlatform/komodo-defi-framework/blob/main/.github/workflows/dev-build.yml#L13
can remove if unnecessary @onur-ozkan

shamardy replied:
Yes, please remove it. Thanks

gcharang replied:
done

onur-ozkan replied:
It can be removed from all scripts, they are outdated. We control those flags with cargo directly since sometime.

gcharang replied:
@onur-ozkan should I remove it from all gh actions in this PR?

shamardy replied:
Yes, please

gcharang replied:
@shamardy done

shamardy commented:
nit: should we add
```yaml
concurrency:
  group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}
  cancel-in-progress: true
  ```
Since it deployed multiple quick times in a row for multiple pushes.

gcharang commented:
@shamardy 

>   cancel-in-progress: true

no, imo 

in this PR, since I was testing, I removed, added label, pushed useless commits to check if it skips/builds etc

in practice, people will add the label once

and every commit after that should get a deployment even if they are pushed with little time in between

shamardy commented:
> @shamardy
> 
> > cancel-in-progress: true
> 
> no, imo
> 
> in this PR, since I was testing, I removed, added label, pushed useless commits to check if it skips/builds etc
> 
> in practice, people will add the label once
> 
> and every commit after that should get a deployment even if they are pushed with little time in between

I don't mind that, since the label shouldn't be added while deving / PR in draft in most cases. So these quick short commits will not be done most of the times.

onur-ozkan commented:
> @shamardy 
> 
> >   cancel-in-progress: true
> 
> no, imo 
> 
> in this PR, since I was testing, I removed, added label, pushed useless commits to check if it skips/builds etc
> 
> in practice, people will add the label once
> 
> and every commit after that should get a deployment even if they are pushed with little time in between
> 

I agree with Omar. If we want to see change between comits, we push and wait until we get the deployment before pushing for other one. In most cases, the reason why we send another commit right after the new commit is because we fix some invalid code. Deploying for previous commits will just waste CI resources for 99% times.

onur-ozkan reviewed (dismissed):
LGTM, thanks!

mariocynicys reviewed (dismissed):
LGTM 🔥

onur-ozkan commented:
Build pipelines are broken currently.

gcharang committed:
implements suggestions, other fixes
